### DISCUSSION TITLE
Learnable Parameters of Data Augmentation Operations

### DISCUSSION THREAD

**Discussion by ArnolFokam** [kornia/kornia#1792] [Questions]
Hi, I am using Kornia in a self-supervised learning task where the objective is to learn augmentation parameters of a simple rotation operation. 

The task is to minimize a reconstruction loss between an image rotated by a predefined degree and the same image rotated by a simple model that learns the parameters of the rotation operation .The idea is to make the model converge to that predefined degree. 

I designed this simple task to see if I could use Kornia to learn augmentation parameters. Here is the rotation model

```python
class RotationModelV2(nn.Module):
    def __init__(self):
        super(RotationModelV2, self).__init__()
        # define the learnable parameter of the rotation operation
        self.Wr = nn.Parameter(torch.rand(1), requires_grad=True)

    def forward(self, image: torch.Tensor) -> torch.Tensor:
        return K.geometry.transform.rotate(image, -360 * self.Wr + 360)
```

However, my loss does not converge as you can see on the graph below. The different colors just means different pre-defined rotation degree to learn for each run.

![W B Chart 7_13_2022, 2_18_38 PM](https://user-images.githubusercontent.com/32764854/178731991-bcae0d2d-7186-46fd-a8ec-aa7f53a05858.png)

# Questions

Therefore, I have the following questions:
- Is there anything missing in my implementation?
- Has anyone done something similar? If yes, I would love to know the outcome.

**Comment by edgarriba** [comment]
Are you tracking the learned rotation values ? and eventually the output of rotated images ?

**Reply by ArnolFokam** [reply]
Yes but the plot above is the history of the reconstruction loss during training. It is basically the absolute difference between the image rotated with a predefined constant angle (target) and the image rotated with the learnable rotation parameter (prediction).

Normally, I would expect the learnable rotation parameter to get more and more closer to the predefined constant angle of rotation during training. It happens for some experiments but this is very rare.

**Reply by edgarriba** [reply]
is it possible to share more code? e.g if you use augmentations, or any other insight

**Reply by ArnolFokam** [reply]
Ok, no problem. Here is the complete training PyTorch Lightning Module

```python
from argparse import Namespace
from typing import Tuple

import numpy as np
from PIL import Image
import torch
from torch import nn
import pytorch_lightning as pl
from torch.utils.data import DataLoader
from torchvision.datasets import CIFAR10
import kornia as K


class RotationModelV2(nn.Module):
    """
    This is the pytorch lightning module that learns how to 
    rotate an input image so that it matches the target image
    """
    
    def __init__(self):
        super(RotationModelV2, self).__init__()
        # we define the learnable rotation parameter
        self.Wr = nn.Parameter(torch.rand(1), requires_grad=True)

    def forward(self, image: torch.Tensor) -> torch.Tensor:
        return K.geometry.transform.rotate(image, -360 * self.Wr + 360)

    @property
    def degree(self):
        return (-360 * self.Wr + 360).item()


class Transform(nn.Module):
    """
    This is the module that acts as a preprocessing 
    step and provides an input and a target image as a 
    tuple of tensors.
    
    Input Image: The input image is just a simple image from the chosen dataset.
    
    Target Image: The target image is the image rotated by a degree. 
                  The degree's value was instantiated during the module's
                  initialization. This value stays constant for every pair of input/target
    """
    def __init__(self) -> None:
        super().__init__()
        
        # pre-defined constant degree value to be learnt but the rotation model.
        self.degree = torch.randint(low=0, high=360, size=(1, )) # set a constant value randomly for each experiment

    @torch.no_grad()  # disable gradients for efficiency
    def forward(self, x: Image) -> Tuple[torch.Tensor, torch.Tensor]:
        x_tmp: np.ndarray = np.array(x)  # HxWxC
        x_out: torch.Tensor = K.image_to_tensor(x_tmp, keepdim=True)  # CxHxW
        x_out = x_out.float()
        return x_out, K.geometry.transform.rotate(x_out.unsqueeze(0), angle=self.degree.float()).squeeze(0)


class SSLModel(pl.LightningModule):

    def __init__(self, args: Namespace):
        super(SSLModel, self).__init__()
        self.save_hyperparameters()
        self.preprocess = Transform()  # preprocessing step that returns (input, target) image
        self.rotator = RotationModelV2()  # model that learns to rotate the image correctly
        self.l1 = nn.L1Loss() # reconstruction loss

    def training_step(self, batch, batch_idx):
        # REQUIRED
        (x, y), _ = batch  # we get the tuple pair and ignore the labels (self-supervised)
        x_aug = self.rotator(x)  # rotate the input image
        y = y.squeeze()
        loss = self.l1(x_aug, y) # calculate absolute difference between preprocessing step's rotation and model's rotation
        self.log('train_loss', loss)
        return loss

    def validation_step(self, batch, batch_idx):
        # OPTIONAL
        (x, y), _ = batch
        return self.l1(self.rotator(x), y.squeeze())

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.args.learning_rate)

    def prepare_data(self):
        CIFAR10(self.hparams.args.data_dir, train=True, download=True, transform=self.preprocess)
        CIFAR10(self.hparams.args.data_dir, train=False, download=True, transform=self.preprocess)

    def train_dataloader(self):
        dataset = CIFAR10(self.hparams.args.data_dir, train=True, download=False, transform=self.preprocess)
        loader = DataLoader(
            dataset,
            batch_size=self.hparams.args.batch_size,
            num_workers=8)
        return loader

    def val_dataloader(self):
        dataset = CIFAR10(self.hparams.args.data_dir, train=False, download=False, transform=self.preprocess)
        loader = DataLoader(
            dataset,
            batch_size=self.hparams.args.batch_size,
            num_workers=8)
        return loader
```

**Reply by edgarriba** [reply]
not sure if you need to clamp the parameter value/grads to avoid overflows. there was a discussion on that in our slack and some suggestions in this direction.

**Reply by ArnolFokam** [reply]
Yes, I was originally using a sigmoid function with a deeper model to ensure the param values stayed in the range [0, 1] (also tried various learning rates).

I would love to know how the param values of each operations are updated w.r.t gradients. My hypothesis is that the way gradient information received is used to update the parameter's value drastically influences the learning process. 

If you know anyone who did something similar and was able to achieve consistent convergence, I would love to chat with him/her.

**Reply by edgarriba** [reply]
join our slack: https://kornia.readthedocs.io/en/latest/#join-the-community

/cc @shijianjian
